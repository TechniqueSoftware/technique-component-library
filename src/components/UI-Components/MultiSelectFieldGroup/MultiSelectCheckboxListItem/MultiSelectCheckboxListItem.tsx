import * as React from 'react';
import { Theme } from '@material-ui/core';
import ListItem from '@material-ui/core/ListItem';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import withStyles from '@material-ui/styles/withStyles';
import { Checkbox, CheckboxProps } from '../../Checkbox/Checkbox';

const styles = (theme: Theme) => ({
  checkbox: {
    padding: theme.spacing(.5)
  }
});

export interface MultiSelectCheckboxListItemProps {
  classes?: any;
  groupId: string;
  name: string;
  labelString: string;
  labelComponent?: React.ReactElement<any>;
  value: string;
  onClick: (eventData: MultiSelectCheckboxListItemClickEventData) => void;
  checked: boolean;
  disabled?: boolean;
  indeterminate?: boolean;
}

export interface MultiSelectCheckboxListItemClickEventData {
  groupId: string;
  value: string;
  checked: boolean;
}

export class MultiSelectCheckboxListItem extends React.Component<MultiSelectCheckboxListItemProps> {

  static defaultProps: Partial<MultiSelectCheckboxListItemProps> = {
    indeterminate: false,
    disabled: false
  };

  constructor(props: any) {
    super(props);
    this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
  }

  shouldComponentUpdate(nextProps: MultiSelectCheckboxListItemProps) {
    return nextProps.checked !== this.props.checked ||
      nextProps.indeterminate !== this.props.indeterminate;
  }

  handleCheckboxClick(event: any) {
    const { groupId, value } = this.props;
    const { target } = event;
    const multiSelectCheckboxListItemClickEventData: MultiSelectCheckboxListItemClickEventData = {
      groupId,
      value,
      checked: target.checked
    };
    this.props.onClick(multiSelectCheckboxListItemClickEventData);
  }

  render() {
    const { name, value, labelString, checked, indeterminate, labelComponent, disabled, classes } = this.props;

    const checkboxProps: CheckboxProps = {
      indeterminate,
      className: classes.checkbox,
      color: 'primary',
      onClick: this.handleCheckboxClick
    };

    return (
      <ListItem key={`${name}-${labelString}`} dense={true}>
        <FormControlLabel
          name={name}
          disabled={disabled}
          label={labelComponent || labelString}
          value={value}
          checked={checked}
          control={<Checkbox {...checkboxProps} />}
        />
      </ListItem>
    );
  }
}

export default withStyles(styles)(MultiSelectCheckboxListItem);
