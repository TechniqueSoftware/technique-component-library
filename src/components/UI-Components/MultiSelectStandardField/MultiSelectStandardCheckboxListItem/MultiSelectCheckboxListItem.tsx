import * as React from 'react';
import { Theme } from '@material-ui/core';
import withStyles from '@material-ui/styles/withStyles';

import FormControlLabel from '@material-ui/core/FormControlLabel';
import ListItem from '@material-ui/core/ListItem';

import { Checkbox, CheckboxProps } from '../../Checkbox/Checkbox';
import { MultiSelectStandardOptions } from '../MultiSelectStandardSearchList/MultiSelectStandardSearchList';

const styles = (theme: Theme) => ({
  checkbox: {
    padding: theme.spacing(.5)
  }
});

export interface MultiSelectCheckboxListItemProps {
  classes?: any;
  name: string;
  labelString: string;
  value: string;
  onClick: (eventData: MultiSelectStandardOptions[]) => void;
  checked: boolean;
  disabled?: boolean;
  multiSelectStandardFieldProps: MultiSelectStandardOptions[];
}

export class MultiSelectCheckboxListItem extends React.Component<MultiSelectCheckboxListItemProps> {
  static defaultProps: Partial<MultiSelectCheckboxListItemProps> = {
    disabled: false
  };

  constructor(props: any) {
    super(props);
    this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
  }

  handleCheckboxClick(event: any) {
    const { multiSelectStandardFieldProps } = this.props;
    const newArr = multiSelectStandardFieldProps.map(obj => {
      if (obj.value === event.target.value) {
        return {
          ...obj, checked: event.target.checked
        };
      }
      return obj;
    });
    this.props.onClick(newArr);
  }

  render() {
    const { name, value, labelString, checked, disabled, classes } = this.props;

    const checkboxProps: CheckboxProps = {
      className: classes.checkbox,
      color: 'primary',
      onClick: this.handleCheckboxClick
    };

    return (
      <ListItem key={`${name}-${labelString}`} dense={true}>
        <FormControlLabel
          name={name}
          disabled={disabled}
          label={labelString}
          value={value}
          checked={checked}
          control={<Checkbox {...checkboxProps} />}
        />
      </ListItem>
    );
  }
}

export default withStyles(styles)(MultiSelectCheckboxListItem);
