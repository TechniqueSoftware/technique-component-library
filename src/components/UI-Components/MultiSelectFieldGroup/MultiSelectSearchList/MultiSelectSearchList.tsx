import * as React from 'react';
import Search from '@material-ui/icons/Search';
import { Theme } from '@material-ui/core';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import withStyles from '@material-ui/styles/withStyles';
import MultiSelectCheckboxListItem, {
  MultiSelectCheckboxListItemClickEventData,
  MultiSelectCheckboxListItemProps
} from '../MultiSelectCheckboxListItem/MultiSelectCheckboxListItem';
import TextInput, { TextInputProps } from '../../TextInput/TextInput';

const debounce = require('lodash.debounce');

const styles = (theme: Theme) => ({
  container: {
    border: `1px solid ${theme.palette.grey['300']}`,
    background: theme.palette.common.white
  },
  whiteBackground: {
    background: theme.palette.common.white
  }
});

export type MultiSelectGroupValueWrapper = { groupId: string, values: Set<string> };

export interface MultiSelectSearchListProps {
  readOnly?: boolean;
  error?: boolean;
  classes?: any;
  multiSelectOptionGroupDataList: MultiSelectOptionGroupData[];
  groupValueWrapperList: MultiSelectGroupValueWrapper[];
  onCheckboxGroupChange: (groupValues: MultiSelectGroupValueWrapper[]) => void;
  listContainerHeight?: number;
}

export interface MultiSelectOptionGroupData {
  groupId: string;
  groupLabel: string;
  multiSelectOptionDataList: MultiSelectOptionData[];
}

export interface MultiSelectOptionData {
  name: string;
  labelString: string;
  labelComponent?: React.ReactElement<any>;
  value: string;
  disabled?: boolean;
}

export interface MultiSelectSearchListState {
  search: string;
  debouncedSearch: string;
  debounceSearchRegExp: RegExp;
  lastCheckboxChangeTimestamp: number;
}

export class MultiSelectSearchList extends React.Component<MultiSelectSearchListProps, MultiSelectSearchListState> {

  static defaultProps: Partial<MultiSelectSearchListProps> = {
    listContainerHeight: 450,
    readOnly: false
  };

  constructor(props: any) {
    super(props);
    this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
    this.handleCheckboxGroupChange = this.handleCheckboxGroupChange.bind(this);
    this.handleSearchChange = this.handleSearchChange.bind(this);
    this.updateLastCheckboxChangeTimestamp = this.updateLastCheckboxChangeTimestamp.bind(this);
    this.updateDebouncedSearch = debounce(this.updateDebouncedSearch, 0);

    this.state = {
      search: '',
      debouncedSearch: '',
      debounceSearchRegExp: new RegExp('', 'i'),
      lastCheckboxChangeTimestamp: new Date().getTime()
    }
    ;
  }

  shouldComponentUpdate(nextProps: MultiSelectSearchListProps, nextState: MultiSelectSearchListState) {
    // Update if the search has changed
    return (
      // If error is changed
      this.props.error !== nextProps.error ||
      // Or if the search has changed
      this.state.debouncedSearch !== nextState.debouncedSearch ||
      // Or if a checkbox change was made
      this.state.lastCheckboxChangeTimestamp !== nextState.lastCheckboxChangeTimestamp ||
      // Or if any of the groups sizes changed
      this.props.groupValueWrapperList.some((group) => {
        const nextGroup = nextProps.groupValueWrapperList.find((nextGroup) => {
          return nextGroup.groupId === group.groupId;
        });
        return nextGroup.values.size !== group.values.size;
      }) ||
      // Or if any of the group option sizes change
      this.props.multiSelectOptionGroupDataList.some((group) => {
        const nextGroup = nextProps.multiSelectOptionGroupDataList.find((nextGroup) => {
          return nextGroup.groupId === group.groupId;
        });
        return nextGroup.multiSelectOptionDataList.length !== group.multiSelectOptionDataList.length;
      })
    );
  }

  updateDebouncedSearch(search: string) {
    this.setState({
      debouncedSearch: search,
      debounceSearchRegExp: new RegExp(search, 'i')
    });
  }

  handleSearchChange(event: any) {
    const search = event.target.value;
    this.setState({ search });
    this.updateDebouncedSearch(search);
  }

  handleCheckboxGroupChange(eventData: MultiSelectCheckboxListItemClickEventData) {
    const { groupValueWrapperList, onCheckboxGroupChange, multiSelectOptionGroupDataList, } = this.props;
    const isGroupMatchingEvent = (group: Partial<{ groupId: string }>) => group.groupId === eventData.groupId;
    const optionGroupData = multiSelectOptionGroupDataList.find(isGroupMatchingEvent);
    const groupValueWrapperListCopy = groupValueWrapperList.map((item) => ({ ...item }));
    const groupValues = groupValueWrapperListCopy.find(isGroupMatchingEvent);
    if (!optionGroupData || !groupValues) {
      return false;
    }
    if (eventData.checked) {
      const allOptions: Set<any> = new Set();
      optionGroupData.multiSelectOptionDataList.forEach((item) => {
        if (!item.disabled) {
          allOptions.add(item.value);
        }
      });
      groupValues.values = allOptions;
    } else {
      groupValues.values.clear();
    }
    onCheckboxGroupChange(groupValueWrapperListCopy);
    this.updateLastCheckboxChangeTimestamp();
  }

  searchIncludes(string: string) {
    if (this.state.debouncedSearch) {
      return this.state.debounceSearchRegExp.test(string);
    }
    // If there is no search query then everything matches
    return true;
  }

  handleCheckboxClick(eventData: MultiSelectCheckboxListItemClickEventData) {
    const { groupValueWrapperList, onCheckboxGroupChange } = this.props;

    const groupValuesCopy: MultiSelectGroupValueWrapper[] = groupValueWrapperList.map((item) => ({ ...item }));
    const group = groupValuesCopy.find((group) => group.groupId === eventData.groupId);
    if (group) {
      if (eventData.checked) {
        group.values.add(eventData.value);
      } else {
        group.values.delete(eventData.value);
      }
    }
    onCheckboxGroupChange(groupValuesCopy);
    this.updateLastCheckboxChangeTimestamp();
  }

  /**
   * We update the timestamp of the last change so we can easily do a compare in shouldComponentUpdate to know when
   * our list has had a change. If we didn't do this we'd have to compare the entire groupValueWrapperList list.
   */
  updateLastCheckboxChangeTimestamp() {
    this.setState({ lastCheckboxChangeTimestamp: new Date().getTime() });
  }

  renderCheckboxListItemWithHeader(multiSelectOptionGroupData: MultiSelectOptionGroupData) {
    const { groupLabel, multiSelectOptionDataList, groupId } = multiSelectOptionGroupData;
    const { groupValueWrapperList, readOnly } = this.props;
    const { debouncedSearch } = this.state;

    const group = groupValueWrapperList.find((group) => group.groupId === groupId) || { groupId, values: new Set() };
    const allSelected = multiSelectOptionDataList.every((optionData) => optionData.disabled || group.values.has(optionData.value));
    const someSelected = multiSelectOptionDataList.some((optionData) => group.values.has(optionData.value));
    const searchMatchesSome = multiSelectOptionGroupData.multiSelectOptionDataList.some((optionData) => {
      return this.searchIncludes(optionData.labelString);
    });

    const headerProps: MultiSelectCheckboxListItemProps = {
      groupId,
      disabled: readOnly,
      labelString: groupLabel,
      checked: allSelected,
      name: `${groupId}-allSelected`,
      value: groupId,
      onClick: this.handleCheckboxGroupChange,
      indeterminate: someSelected && !allSelected
    };

    if (!searchMatchesSome) return null;
    return (
      <React.Fragment key={`checkboxList-${groupLabel}`}>
        <MultiSelectCheckboxListItem {...headerProps} />
        <ul>
          {multiSelectOptionDataList.map(multiSelectOptionData => {
            const { name, labelString, value, labelComponent, disabled } = multiSelectOptionData;
            const multiSelectCheckboxListItemProps: MultiSelectCheckboxListItemProps = {
              groupId,
              labelString,
              labelComponent,
              value,
              name,
              disabled: readOnly || disabled,
              checked: group.values.has(value),
              onClick: this.handleCheckboxClick
            };
            if (debouncedSearch && !this.searchIncludes(labelString)) return null;
            return (<MultiSelectCheckboxListItem {...multiSelectCheckboxListItemProps} key={`${groupId}-${value}`} />);
          })}
        </ul>
      </React.Fragment>);
  }

  renderSearchInput() {
    const { error = false } = this.props;
    const { search } = this.state;
    const label = 'Search';
    const props: TextInputProps = {
      error,
      value: search,
      'aria-label': label,
      placeholder: label,
      startAdornmentIcon: <Search />,
      onChange: this.handleSearchChange,
    };
    return (<TextInput {...props} />);
  }

  renderListItems() {
    return this.props.multiSelectOptionGroupDataList.map(
      (multiSelectOptionGroupData) => {
        return this.renderCheckboxListItemWithHeader(multiSelectOptionGroupData);
      });
  }

  render() {
    const { search } = this.state;
    const { multiSelectOptionGroupDataList, classes, listContainerHeight } = this.props;

    let searchMatchesSome;
    if (search) {
      searchMatchesSome = multiSelectOptionGroupDataList.some((optionGroupData) => {
        return optionGroupData.multiSelectOptionDataList.some((optionData) => {
          return this.searchIncludes(optionData.labelString);
        });
      });
    }

    const listStyles = { overflow: 'auto', maxHeight: listContainerHeight, width: '100%' };

    return (
      <div className={classes.container}>
        {this.renderSearchInput()}
        <List subheader={<li />} style={listStyles} dense>
          {(!search || search && searchMatchesSome) && this.renderListItems()}
          {search && !searchMatchesSome && <ListItem><ListItemText>No match found</ListItemText></ListItem>}
        </List>
      </div>);
  }
}

export default withStyles(styles)(MultiSelectSearchList);
