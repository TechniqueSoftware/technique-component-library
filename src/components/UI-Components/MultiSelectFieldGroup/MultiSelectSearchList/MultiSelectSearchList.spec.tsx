import { shallow } from 'enzyme';
import * as React from 'react';
import { MultiSelectSearchList, MultiSelectSearchListProps } from './MultiSelectSearchList';
import { MultiSelectCheckboxListItemClickEventData } from '../MultiSelectCheckboxListItem/MultiSelectCheckboxListItem';

describe('MultiSelectSearchList', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  const baseProps: MultiSelectSearchListProps = {
    classes: {
      searchContainer: 'mockSearchContainerClass'
    },
    onCheckboxGroupChange: jest.fn(),
    groupValueWrapperList: [{ values: new Set(['1', '2', '3']), groupId: 'group1' }],
    multiSelectOptionGroupDataList: [{
      groupId: 'group1',
      groupLabel: 'Group 1',
      multiSelectOptionDataList: [{
        labelComponent: undefined,
        name: 'mockLabelName',
        value: '1',
        labelString: 'group1-mockLabelText1'
      }, {
        labelComponent: undefined,
        name: 'mockLabelName2',
        value: '2',
        labelString: 'group1-mockLabelText2'
      }
      ]
    },
      {
        groupId: 'group2',
        groupLabel: 'Group 2',
        multiSelectOptionDataList: [{
          labelComponent: undefined,
          name: 'mockLabelName',
          value: '3',
          labelString: 'group2-mockLabelText1'
        }, {
          labelComponent: undefined,
          name: 'mockLabelName2',
          value: '4',
          labelString: 'group2-mockLabelText2'
        }
        ]
      }]
  };

  test('should not show items when search does not match and should show "No Match Found" message', () => {
    const props: MultiSelectSearchListProps = {
      ...baseProps,
    };
    const shallowWrapper = shallow(<MultiSelectSearchList {...props} />);
    const searchInputWrapper = shallowWrapper.find('WithStyles(TextInput)');
    searchInputWrapper.simulate('change', { target: { value: 'xxx' } });
    jest.runAllTimers();
    shallowWrapper.instance().forceUpdate();
    expect(shallowWrapper.debug()).toMatchSnapshot();
    const listWrapper = shallowWrapper.find('WithStyles(List)');
    const debugView = listWrapper.debug();
    expect(debugView.includes('No match found'));
    expect(debugView).toMatchSnapshot();
  });

  test('should only show items that match searchText when search text is entered', () => {
    const props: MultiSelectSearchListProps = {
      ...baseProps,
    };
    const shallowWrapper = shallow(<MultiSelectSearchList {...props} />);

    const multiSelectOptionGroupDatum = baseProps.multiSelectOptionGroupDataList[0];
    const firstItemLabelText = multiSelectOptionGroupDatum.multiSelectOptionDataList[0].labelString;
    const searchInputWrapper = shallowWrapper.find('WithStyles(TextInput)');
    expect(searchInputWrapper.length).toBe(1);
    searchInputWrapper.simulate('change', { target: { value: firstItemLabelText } });
    // Search is debounced so we need to run timers
    jest.runAllTimers();
    shallowWrapper.instance().forceUpdate();
    const listWrapper = shallowWrapper.find('WithStyles(ForwardRef(List))');
    const checkboxListItem = listWrapper.find('WithStyles(MultiSelectCheckboxListItem)');

    expect(listWrapper.debug()).toMatchSnapshot();
    expect(listWrapper.debug().includes(firstItemLabelText));
    expect(listWrapper.debug().includes(multiSelectOptionGroupDatum.groupLabel));
    // 2 due to group checkbox
    expect(checkboxListItem.length).toBe(2);
  });

  test('checking a checkbox group header should return a list of all items in the group when checked is true', () => {
    const props: MultiSelectSearchListProps = {
      ...baseProps,
      onCheckboxGroupChange: jest.fn(),
    };
    const shallowWrapper = shallow(<MultiSelectSearchList {...props} />);

    const firstGroup = props.multiSelectOptionGroupDataList[0];
    const firstGroupCheckbox = shallowWrapper.findWhere((wrapper) => wrapper.prop('labelString') === firstGroup.groupLabel);

    const eventData: MultiSelectCheckboxListItemClickEventData = {
      groupId: firstGroup.groupId,
      checked: true,
      value: ''
    };

    firstGroupCheckbox.simulate('click', eventData);

    const onCheckboxGroupChange: any = props.onCheckboxGroupChange;
    expect(onCheckboxGroupChange.mock.calls[0][0][0]).toEqual({
      groupId: firstGroup.groupId,
      values: new Set(firstGroup.multiSelectOptionDataList.map((value) => value.value))
    });
  });

  test('checking a checkbox group header should return an empty set when checked is false', () => {
    const props: MultiSelectSearchListProps = {
      ...baseProps,
      onCheckboxGroupChange: jest.fn(),
    };
    const shallowWrapper = shallow(<MultiSelectSearchList {...props} />);

    const firstGroup = props.multiSelectOptionGroupDataList[0];
    const firstGroupCheckbox = shallowWrapper.findWhere((wrapper) => wrapper.prop('labelString') === firstGroup.groupLabel);

    const eventData: MultiSelectCheckboxListItemClickEventData = {
      groupId: firstGroup.groupId,
      checked: false,
      value: ''
    };

    firstGroupCheckbox.simulate('click', eventData);

    const onCheckboxGroupChange: any = props.onCheckboxGroupChange;
    expect(onCheckboxGroupChange.mock.calls[0][0][0]).toEqual({
      groupId: firstGroup.groupId,
      values: new Set()
    });
  });
});
