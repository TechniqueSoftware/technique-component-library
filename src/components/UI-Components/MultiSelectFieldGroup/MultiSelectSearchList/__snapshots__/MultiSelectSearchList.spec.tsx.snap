// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectSearchList should not show items when search does not match and should show "No Match Found" message 1`] = `
"<div className={[undefined]}>
  <WithStyles(TextInput) error={false} value=\\"xxx\\" aria-label=\\"Search\\" placeholder=\\"Search\\" startAdornmentIcon={{...}} onChange={[Function: bound ]} />
  <WithStyles(ForwardRef(List)) subheader={{...}} style={{...}} dense={true}>
    <WithStyles(ForwardRef(ListItem))>
      <WithStyles(ForwardRef(ListItemText))>
        No match found
      </WithStyles(ForwardRef(ListItemText))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</div>"
`;

exports[`MultiSelectSearchList should not show items when search does not match and should show "No Match Found" message 2`] = `""`;

exports[`MultiSelectSearchList should only show items that match searchText when search text is entered 1`] = `
"<WithStyles(ForwardRef(List)) subheader={{...}} style={{...}} dense={true}>
  <WithStyles(MultiSelectCheckboxListItem) groupId=\\"group1\\" disabled={false} labelString=\\"Group 1\\" checked={true} name=\\"group1-allSelected\\" value=\\"group1\\" onClick={[Function: bound ]} indeterminate={false} />
  <ul>
    <WithStyles(MultiSelectCheckboxListItem) groupId=\\"group1\\" labelString=\\"group1-mockLabelText1\\" labelComponent={[undefined]} value=\\"1\\" name=\\"mockLabelName\\" disabled={[undefined]} checked={true} onClick={[Function: bound ]} />
  </ul>
</WithStyles(ForwardRef(List))>"
`;
